### PR TITLE
feat: Add vite_picture_tag for Rails 7.1

### LINKED ISSUE #386: Feature request: vite_picture_tag
**Is your feature request related to a problem? Please describe.**

Rails [introduced](https://github.com/rails/rails/pull/48100) a `picture_tag` helper earlier this year. It is on the main branch (not yet released) and presumably will be part of Rails 7.1.

**Describe the solution you'd like**

I'd like to see a `vite_picture_tag` wrapper, perhaps with an implementation like this:

```ruby
def vite_picture_tag(*sources, &block)
  sources.flatten!
  options = sources.extract_options!
  vite_sources = sources.map { |src| vite_asset_path(src) }
  
  # Delegate to ActionView::Helpers::AssetTagHelper#picture_tag
  picture_tag(*vite_sources, options, &block)
end
```

**Describe alternatives you've considered**

The tag helper is syntax sugar; `<picture>` tags can of course still be created manually without too much effort:

```ruby
tag.picture do
  tag.source srcset: vite_asset_path("images/one.png")
  tag.source srcset: vite_asset_path("images/two.png")
  vite_image_tag("images/two.png", alt: "example")
end
```

Versus:

```ruby
vite_picture_tag("images/one.png", "images/two.png", image: {alt: "example"})
```

**Additional context**

I'd be happy to contribute a PR, but I am wondering: should I wait until Rails 7.1 is released?

### PR DESCRIPTION
**Pull request by mattbrictson**
### Description 📖

This commit adds a `vite_picture_tag` helper that handles resolving assets via Vite, but otherwise behaves the same as the Rails built-in `picture_tag`. It is only exposed for Rails >= 7.1.0. Fixes #386 

### Background 📜

Rails 7.1 introduced a `picture_tag` helper.[^1] 

[^1]: https://github.com/rails/rails/pull/48100

### The Fix 🔨

Added `vite_picture_tag` to `ViteRails::TagHelpers` with a corresponding test.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -22,188 +22,240 @@ PATH
 GEM
   remote: https://rubygems.org/
   specs:
-    actioncable (6.1.3.1)
-      actionpack (= 6.1.3.1)
-      activesupport (= 6.1.3.1)
+    actioncable (7.1.0)
```
Comment (mattbrictson) [inline Gemfile.lock:25]: 🗒️ I updated the lock file to be able to test against Rails 7.1.0. Let me know if I should split this into a separate commit or PR.

File: vite_rails/lib/vite_rails/tag_helpers.rb [hunk 2]
```diff
@@ -79,6 +79,17 @@ def vite_image_tag(name, **options)
     image_tag(vite_asset_path(name), options)
   end
 
+  if defined?(Rails) && Rails.gem_version >= Gem::Version.new('7.1.0')
```
[the remark this answers is not in the document]
Comment (ElMassimo) [reply vite_rails/lib/vite_rails/tag_helpers.rb]: That approach would have worked, but I'd like to [provide more information](https://github.com/ElMassimo/vite_ruby/pull/409/commits/f517b49a8fd8e220f8981cb265033378ec86db77) in case users see it in the docs and try to use it in an older version of Rails.

**Comment by ElMassimo** [comment]
Nice addition, thanks Matt!

File: test/helper_test.rb [hunk 3]
```diff
@@ -180,4 +180,16 @@ def test_vite_image_tag
         vite_image_tag('images/logo.png', srcset: { 'images/logo-2x.png' => '2x' }, alt: 'Logo')
     }
   end
+
+  if Rails.gem_version >= Gem::Version.new('7.1.0')
+    def test_vite_picture_tag
+      assert_equal <<~HTML.gsub(/\n\s*/, ''), vite_picture_tag('images/logo.svg', 'images/logo.png', class: 'test', image: { alt: 'Logo' })
+        <picture class="test">
+          <source srcset="/vite-production/assets/logo.322aae0c.svg" type="image/svg+xml" />
```
Comment (mjankowski) [inline test/helper_test.rb:188]: I realize this PR is ~9 months old, but this line has different output for me.

Running Rails 7.1.3.4, the `type` attribute is different. I see a failure like:

```
Failure:
HelperTest#test_vite_picture_tag [/Users/mjankowski/repos/vite_ruby/test/helper_test.rb:186]
Minitest::Assertion: --- expected
+++ actual
@@ -1 +1 @@
-"<picture class=\"test\"><source srcset=\"/vite-production/assets/logo.322aae0c.svg\" type=\"image/svg+xml\" /><source srcset=\"/vite-production/assets/logo.f42fb7ea.png\" type=\"image/png\" /><img alt=\"Logo\" src=\"/vite-production/assets/logo.f42fb7ea.png\" /></picture>"
+"<picture class=\"test\"><source srcset=\"/vite-production/assets/logo.322aae0c.svg\" type=\"svg\" /><source srcset=\"/vite-production/assets/logo.f42fb7ea.png\" type=\"png\" /><img alt=\"Logo\" src=\"/vite-production/assets/logo.f42fb7ea.png\" /></picture>"
```

Maybe related to https://github.com/rails/rails/pull/48266 ?

Curious to see a CI run after https://github.com/ElMassimo/vite_ruby/pull/467

Comment (mattbrictson) [reply test/helper_test.rb:188]: @mjankowski I can't reproduce what you are seeing locally. Using vite_ruby @ main, and running `bundle up` to get the latest gems, the test still passes for me:

```
$ bundle exec rake test TEST=test/helper_test.rb
...
14 tests, 36 assertions, 0 failures, 0 errors, 0 skips
```

```
$ bundle show rails
/Users/mbrictson/.rbenv/versions/3.1.1/lib/ruby/gems/3.1.0/gems/rails-7.1.3.4
```

---

Digging through the Rails source, my guess is that `ActionView::Template::SimpleType` is somehow being used in your environment, causing the mime type lookup to return `svg` instead of the correct `image/svg+xml`.

I found this comment:

> SimpleType is mostly just a stub implementation for when Action View is used without Action Dispatch.

So maybe Action Dispatch isn't being loading in your environment, for some reason?

https://github.com/rails/rails/blob/v7.1.3.4/actionview/lib/action_view/template/types.rb#L9

Comment (mjankowski) [reply test/helper_test.rb:188]: Ah, thanks ... think I tracked it down.

I have a branch going trying to get rid of some of deprecation noise during spec run, and I did a bit of branch jam-up on myself, and I think I had `eager_load` disabled during that failure, which may explain the lack of the template being present (and it falling back to just ext string) during the run.
